⚠️ DynamicRestMapper: return NoMatchError when resource doesn't exist

timebertt opened the pull request:
<!-- please add an icon to the title of this PR (see VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠ (:warning:, major), ✨ (:sparkles, minor), 🐛 (:bug:, patch), 📖 (:book:, docs), or 🌱 (:seedling:, other) -->

<!-- What does this do, and why do we need it? -->

Fixes #1120 

This PR fixes the DynamicRESTMapper's behaviour when it is rate limited:
Instead of returning `ErrRateLimited`, it now directly returns the `meta.NoKindMatchError` / `meta.NoResourceMatchError` given by the underlying mapper. This way, existing controller code using `meta.IsNoMatchError` works as expected also in the case when the RESTMapper is rate limited.

**This is breaking ⚠️  because:**
- the `DynamicRESTMapper` does now return a `NoKindMatchError`/`NoResourceMatchError` instead of `ErrRateLimited` in case the resource doesn't exist and it was rate limited
- `ErrRateLimited` and `DelayIfRateLimited` are removed, as they aren't needed anymore

**Additional notes:**
The given rate limit can now be seen as an agreement on how often the mapper should refresh its cache (at max). So if the limit is reached, it basically means, we have refreshed often enough (met the contract) and we can safely rely on the cache and directly return the error.

I improved the tests a bit along the way, to ensure all cases are properly tested and made them a bit faster.

timebertt commented on pkg/client/apiutil/dynamicrestmapper.go:
```diff
@@ -29,34 +28,12 @@ import (
-// DelayIfRateLimited returns the delay time until the next API call is
-// allowed and true if err is of type ErrRateLimited. The zero
-// time.Duration value and false are returned if err is not a ErrRateLimited.
-func DelayIfRateLimited(err error) (time.Duration, bool) {
```
We can also think about deprecating `DelayIfRateLimited` (and always returning `0, false`) instead of removing it.
This would make this PR 🐛 instead of ⚠️ FMPOV, so we could cherry-pick it to `release-0.x`.
WDYT?

DirectXMan12 replied:
could also do a wrapper error that wraps both errors if it's a nokindmatcherror that's *caused* by rate limitting.  This has the downside of not working with `meta.IsNoMatchError` (but that function is actually not really correct in Go 1.13+) until they're updated to work with Go's 1.13 standard `errors` package, but has the advantage that it works with idiomatic Go:

```go
if errors.As(err, &meta.NoKindMatchError{}) {
  // no kind matched
}
```

as well as still allowing you to have `DelayIfRateLimited` work.

I'd imagine the implementation would look like:

```go
type ErrorNoMatch struct {
  Delay time.Duration
  matchError error
}

func (e *ErrNoMatch) Unwrap()  error {
  return e.matchError
}
// implement Error() as necessary

// alias for backwards compat

type ErrRateLimited = ErrNoMatch
```

WDYT?

timebertt replied:
Thanks @DirectXMan12 for your suggestion. I like it wrt using go errors idiomatically and also keeping backwards compatability.

Although I am not voting totally against it, I want to note something:
TBH I don't see a real benefit of annotating the `NoKindMatchError` with a delay. I can't think of a use case where a controller/consumer would actually want to distinguish between a "normal" nokindmatch and a nokindmatch that might have been returned by a slightly outdated cache, and then wait until it can retry.
If you added a new CRD for example and you want to add a CR, then you will have to expect a slight delay until the resource comes available in the discovery API anyways, right?
Also, IISIC the code actually doesn't guarantee the caller, that they won't be rate limited again after waiting the returned delay (reservation is cancelled and other callers trying to rediscover might sneak in).

I would think of the specified rate limit as some kind of a SLA between the controller/consumer and the RESTMapper, which says "Please make sure, my discovery cache is not outdated by more than 1s" (if you specify a rate limiter with 1rps).
And if you're call was rate limited because some other rediscoveries exhausted the rate limited only a few milliseconds before yours, than your cache is very likely not to be outdated, although the rate limiter didn't allow the request.

It quite unfortunate, that upstream has not adapted go1.13 errors, yet. But I would expect that an implementation of the `meta.RESTMapper` interface is also working as expected with the provided helper functions in the `meta` package, like `meta.IsNoMatchError`.


As said, I'm not against implementing your approach, I just wanted to give some more background on the motivation why I started with the proposed implementation.

timebertt commented:
/assign @droot

timebertt commented:
/assign @DirectXMan12

DirectXMan12 reviewed (commented):
like the general concept.  Left a comment re: your question.

timebertt commented:
@DirectXMan12 WDYT about my comment?

alvaroaleman commented:
@timebertt can we retitle the PR to clarify the problem its solving rather than implementation details (something like "DynamicRestMapper: Return NoMatchError when resource type doesn't exist") ?

timebertt commented:
/title DynamicRestMapper: return NoMatchError when resource doesn't exist

Any thought on the compatability story, @alvaroaleman?

timebertt commented:
Pinging @alvaroaleman @DirectXMan12 
Any additional feedback on this PR?

timebertt commented:
I made this PR backwards-compatible by deprecating `ErrRateLimited` and associated funcs instead of just removing it.

Can we now proceed with the PR?
/assign @DirectXMan12 @alvaroaleman

vincepri commented on pkg/client/apiutil/dynamicrestmapper.go:
```diff
@@ -31,6 +31,9 @@ import (
 
 // ErrRateLimited is returned by a RESTMapper method if the number of API
 // calls has exceeded a limit within a certain time period.
+// Deprecated: as the dynamic RESTMapper now returns NoMatch errors directly
```
```suggestion
// calls has exceeded a limit within a certain time period.
//
// DEPRECATED: As the dynamic RESTMapper now returns NoMatch errors directly
```

Just to give more emphasis :)

vincepri replied:
Given that we're working on v0.7.x on the main branch, can we actually just remove this to avoid confusion? 

cc @alvaroaleman @DirectXMan12

timebertt replied:
> Just to give more emphasis :)

`Deprecated:` is actually convention and some tools test for this explicitly (ref https://github.com/golang/go/wiki/Deprecated), so I would keep it, if you don't mind :)

timebertt replied:
> Given that we're working on v0.7.x on the main branch, can we actually just remove this to avoid confusion?

Sure, I'm open for it. You decide, what better fits you :)

vincepri replied:
> Deprecated: is actually convention and some tools test for this explicitly (ref https://github.com/golang/go/wiki/Deprecated), so I would keep it, if you don't mind :)

SGTM, let's add a space though between the deprecation notice and the original comment, if we're going down this path.

> Sure, I'm open for it. You decide, what better fits you :)

I'm +1 for it, although I'd like to wait to hear from other folks

timebertt replied:
One additional option would be to keep this PR backwards-compatible as it is, so we can cherry-pick it to `0.6` and `0.5`, and remove the deprecated types/funcs in a separate PR only targeting `>= 0.7`.

vincepri replied:
Hm, not sure if we want to backport these changes, given that's a breaking behavior change (check one error vs another) and could impact some users

timebertt replied:
Yeah, fair enough...

alvaroaleman replied:
This PR is never going to be backwards compatible even if we keep the error type because that error type is never returned (which is btw much more subtile and annoying to debug than a compilation error). So I am very much in favor of just removing it.

vincepri commented on pkg/client/apiutil/dynamicrestmapper.go:
```diff
@@ -43,6 +46,9 @@ func (e ErrRateLimited) Error() string {
 // DelayIfRateLimited returns the delay time until the next API call is
 // allowed and true if err is of type ErrRateLimited. The zero
 // time.Duration value and false are returned if err is not a ErrRateLimited.
+// Deprecated: as the dynamic RESTMapper now returns NoMatch errors directly
```
```suggestion
// time.Duration value and false are returned if err is not a ErrRateLimited.
//
// DEPRECATED: As the dynamic RESTMapper now returns NoMatch errors directly
```

vincepri reviewed (commented):
/ok-to-test

vincepri commented:
/milestone v0.7.x

alvaroaleman commented on pkg/client/apiutil/dynamicrestmapper_test.go:
```diff
@@ -57,53 +58,49 @@ var _ = Describe("Dynamic REST Mapper", func() {
-			Expect(callWithOther()).To(Succeed())
-			Expect(callWithTarget()).To(Succeed())
-		})
+			Expect(callWithOther()).To(beNoMatchError())
```
Why does this fail on the first attempt now and only works on the second?

timebertt replied:
It was also failing on the first attempt before (see [L62](https://github.com/kubernetes-sigs/controller-runtime/pull/1151/files/93a54481b9d2acb5231e6fd33663b1f9acf1ad4b#diff-515d07e6da6fc55b48a1766c7948ba61L62)) as the `secondGVK` is not present by default (see [L33](https://github.com/kubernetes-sigs/controller-runtime/pull/1151/files/93a54481b9d2acb5231e6fd33663b1f9acf1ad4b#diff-515d07e6da6fc55b48a1766c7948ba61L33))
I just moved down the first `callWithOther` and added `count` to ensure, that the RESTMapper really tries to rediscover for the call with the unknown resource.

alvaroaleman commented on pkg/client/apiutil/dynamicrestmapper.go:
```diff
@@ -31,6 +31,10 @@ import (
+// Deprecated: as the dynamic RESTMapper now returns NoMatch errors directly
+// instead of RateLimited errors, it is meaningless and only kept for the sake
+// of backwards-compatibility.
 type ErrRateLimited struct {
```
Does it really make sense to keep this and the `DelayIfRateLimited` if we don't use it anymore? The only use case I can see for those is to handle ratelimit responses from the mapper but that won't work anymore, so removing them and make ppls code fail to compile if they use it seems the better option to me here. WDYT?

vincepri replied:
See https://github.com/kubernetes-sigs/controller-runtime/pull/1151#discussion_r495996649 :)

timebertt commented:
/retest

timebertt commented:
@alvaroaleman @vincepri Thanks for your feedback. Makes sense to me.
I removed `ErrRateLimited` and `DelayIfRateLimited`, marked the PR/commit as ⚠️ and added a note to the PR description.
PTAL :)

timebertt commented:
How can I make `pull-controller-runtime-apidiff-master` happy?

alvaroaleman commented:
> How can I make pull-controller-runtime-apidiff-master happy?

You can't, because its a breaking change. The job is optional though.

/lgtm
/approve

Thanks for your work!

timebertt commented:
> > How can I make pull-controller-runtime-apidiff-master happy?
> 
> You can't, because its a breaking change. The job is optional though.

Ah, got it.
Thanks :)
